### ISSUE TITLE
AnsibleUndefinedVariable: 'list object' has no attribute 'ipv4'

### ISSUE THREAD

**Issue by kevin1** [trailofbits/algo#163]
### OS / Environment
Ubuntu 16.04.1 LTS 4.4.0-53-generic
Freshly created by DigitalOcean in the nyc3 region

### Ansible version
ansible 2.2.0.0

### Version of components from `requirements.txt`
ansible 2.2.0.0
dopy 0.3.5
boto 2.44.0
azure 2.0.0rc5
apache-libcloud 1.4.0
six 1.10.0

### Summary of the problem
When I run `./algo`, I get an error in the step "Setup the ipsec.conf file from our template." 

AnsibleUndefinedVariable: 'list object' has no attribute 'ipv4'

After rolling back to a previous commit, 759807a3b62dbd4bb7500968c36c3ec0c5057315, everything works fine.

### Steps to reproduce the behavior
Run the following script, which is based on the instructions from README:
```
apt install -y python-setuptools build-essential libssl-dev libffi-dev python-dev
git clone https://github.com/trailofbits/algo.git && cd algo
git reset --hard bb90bb26a62ca3c3d775210b35311df2693c0095
easy_install pip && pip install -r requirements.txt
cat <<END > cmd
5
localhost

104.236.77.74
y
n
n
y
y
n
n
n
END
./algo < cmd
```

### The way of deployment (cloud or local)
local

### Expected behavior
Installs successfully.

### Actual behavior
```
TASK [vpn : Setup the strongswan.conf file from our template] ******************
ok: [localhost]

TASK [vpn : Setup the ipsec.conf file from our template] ***********************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'list object' has no attribute 'ipv4'"}

PLAY RECAP *********************************************************************
localhost                  : ok=38   changed=6    unreachable=0    failed=1   
```

### Full log
```
root@algo-nyc:~/algo# ./algo

  What provider would you like to use?
    1. DigitalOcean
    2. Amazon EC2
    3. Google Compute Engine
    4. Microsoft Azure
    5. Install to existing Ubuntu server

Enter the number of your desired provider
: 5

Enter IP address of your server: (use localhost for local installation)
: localhost

What user should we use to login on the server? (ignore if you're deploying to localhost)
[root]: 

Enter the public IP address of your server: (IMPORTANT! This IP is used to verify the certificate)
: 104.236.77.74 

Do you want to apply security enhancements?
[y/N]: y

Do you want to install an HTTP proxy to block ads and decrease traffic usage while surfing?
[y/N]: n

Do you want to install a local DNS resolver to block ads while surfing?
[y/N]: n

Do you want to use auditd for security monitoring (see config.cfg)?
[y/N]: y

Do you want each user to have their own account for SSH tunneling?
[y/N]: y

Do you want to enable VPN always when connected to Wi-Fi?
[y/N]: n

Do you want to enable VPN always when connected to the cellular network?
[y/N]: n

Do you want to enable VPN for Windows 10 clients? (Will use insecure algorithms and ciphers)
[y/N]: n

PLAY [Configure the server and install required software] **********************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [local : Add the instance to an inventory group] **************************
skipping: [localhost]

TASK [local : Add the instance to an inventory group] **************************
changed: [localhost]

TASK [local : Waiting for SSH to become available] *****************************
skipping: [localhost]

PLAY [Post-provisioning tasks] *************************************************

TASK [Install prerequisites] ***************************************************
changed: [localhost]

TASK [common : Gather Facts] ***************************************************
ok: [localhost]

TASK [common : Install tools] **************************************************
ok: [localhost] => (item=[u'git', u'screen', u'apparmor-utils', u'uuid-runtime', u'coreutils', u'sendmail', u'iptables-persistent', u'cgroup-tools'])

TASK [common : Loopback for services configured] *******************************
ok: [localhost]

TASK [common : Loopback included into the network config] **********************
ok: [localhost]

TASK [common : Enable packet forwarding for IPv4] ******************************
ok: [localhost] => (item=net.ipv4.ip_forward)
ok: [localhost] => (item=net.ipv4.conf.all.forwarding)

TASK [common : Enable packet forwarding for IPv6] ******************************
ok: [localhost]

TASK [security : Install tools] ************************************************
ok: [localhost] => (item=[u'unattended-upgrades'])

TASK [security : Configure unattended-upgrades] ********************************
ok: [localhost]

TASK [security : Periodic upgrades configured] *********************************
ok: [localhost]

TASK [security : Find directories for minimizing access] ***********************
ok: [localhost] => (item=/usr/local/sbin)
ok: [localhost] => (item=/usr/local/bin)
ok: [localhost] => (item=/usr/sbin)
ok: [localhost] => (item=/usr/bin)
ok: [localhost] => (item=/sbin)
ok: [localhost] => (item=/bin)

TASK [security : Minimize access] **********************************************
ok: [localhost] => (item=(censored due to no_log))
ok: [localhost] => (item=(censored due to no_log))
ok: [localhost] => (item=(censored due to no_log))
ok: [localhost] => (item=(censored due to no_log))
ok: [localhost] => (item=(censored due to no_log))
ok: [localhost] => (item=(censored due to no_log))

TASK [security : Change shadow ownership to root and mode to 0600] *************
ok: [localhost]

TASK [security : change su-binary to only be accessible to user and group root] 
ok: [localhost]

TASK [security : Collect Use of privileged commands] ***************************
changed: [localhost]

TASK [security : Restrict core dumps (with PAM)] *******************************
ok: [localhost]

TASK [security : Restrict core dumps (with sysctl)] ****************************
ok: [localhost]

TASK [security : Disable Source Routed Packet Acceptance] **********************
ok: [localhost] => (item=net.ipv4.conf.all.accept_source_route)
ok: [localhost] => (item=net.ipv4.conf.default.accept_source_route)

TASK [security : Disable ICMP Redirect Acceptance] *****************************
ok: [localhost] => (item=net.ipv4.conf.all.accept_redirects)
ok: [localhost] => (item=net.ipv4.conf.default.accept_redirects)

TASK [security : Disable Secure ICMP Redirect Acceptance] **********************
ok: [localhost] => (item=net.ipv4.conf.all.secure_redirects)
ok: [localhost] => (item=net.ipv4.conf.default.secure_redirects)

TASK [security : Enable Bad Error Message Protection] **************************
ok: [localhost]

TASK [security : Enable RFC-recommended Source Route Validation] ***************
ok: [localhost] => (item=net.ipv4.conf.all.rp_filter)
ok: [localhost] => (item=net.ipv4.conf.default.rp_filter)

TASK [security : Do not send ICMP redirects (we are not a router)] *************
ok: [localhost]

TASK [security : SSH config] ***************************************************
ok: [localhost]

TASK [vpn : Gather Facts] ******************************************************
ok: [localhost]

TASK [vpn : Install StrongSwan] ************************************************
changed: [localhost]

TASK [vpn : Enforcing ipsec with apparmor] *************************************
changed: [localhost] => (item=/usr/lib/ipsec/charon)
changed: [localhost] => (item=/usr/lib/ipsec/lookip)
changed: [localhost] => (item=/usr/lib/ipsec/stroke)

TASK [vpn : Enable services] ***************************************************
ok: [localhost] => (item=apparmor)
ok: [localhost] => (item=strongswan)
ok: [localhost] => (item=netfilter-persistent)

TASK [vpn : Configure iptables so IPSec traffic can traverse the tunnel] *******
ok: [localhost]

TASK [vpn : Configure ip6tables so IPSec traffic can traverse the tunnel] ******
ok: [localhost]

TASK [vpn : Ensure that the strongswan group exist] ****************************
ok: [localhost]

TASK [vpn : Ensure that the strongswan user exist] *****************************
ok: [localhost]

TASK [vpn : Ensure that the strongswan service directory exist] ****************
ok: [localhost]

TASK [vpn : Setup the cgroup limitations for the ipsec daemon] *****************
ok: [localhost]

RUNNING HANDLER [vpn : restart apparmor] ***************************************
changed: [localhost]

TASK [vpn : Setup the strongswan.conf file from our template] ******************
ok: [localhost]

TASK [vpn : Setup the ipsec.conf file from our template] ***********************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'list object' has no attribute 'ipv4'"}

PLAY RECAP *********************************************************************
localhost                  : ok=38   changed=6    unreachable=0    failed=1   
```

**Comment by gunph1ld** [comment]
Can not reproduce, works well on my side

**Comment by dguido** [comment]
@kevin1 Can you try this again? I don't have a spare Ubuntu server to test this on right now.

**Comment by kennwhite** [comment]
FWIW, this seems to be resolved in the current 1483116  master (with PR 179 patch: https://github.com/trailofbits/algo/pull/179/commits/e0712965dc8f5acd367d8906dc64a88ce655c267#diff-412f0e93898687e6289540da55f8861b).
Following the identical config as @kevin1 describes above, I could not reproduce this crash on new Digital Ocean stock Ubuntu instances, even after launching all four combinations of "Private Network" and/or "IPv6" in the DO console. Successful build log here:
https://gist.github.com/kennwhite/c47c7c699efd6d51ed6a66ce622f98ad

**Comment by kevin1** [comment]
I played with this some more and figured out the problem. I had written a cloud-init script that overwrote the config with an old version. Sometime between 759807a and the master at the time of filing the bug, dns_servers changed from:

```
dns_servers:
  - 8.8.8.8
  - 8.8.4.4
  - 2001:4860:4860::8888
  - 2001:4860:4860::8844
```

to:


```
dns_servers:
  ipv4:
    - 8.8.8.8
    - 8.8.4.4
  ipv6:
    - 2001:4860:4860::8888
    - 2001:4860:4860::8844
```

Sorry for the churn! This was a really silly mistake on my end :(
